### ISSUE TITLE
Environment support for SVN

### ISSUE THREAD

**Issue by adrienthebo** [puppetlabs/r10k#104]
SVN doesn't have a fully native concept of branches or tags, but instead relies on a layout convention with the top level directories `trunk/`, `tags/`, and `branches/`. R10k should build an opinionated workflow on top of this given structure to provide a workflow similar to git dynamic environments.

Branch enumeration should be done by doing a `svn checkout` of the top level repository, enumerating `branches/` and mapping each one to an environment, and mapping `trunk/` to production. After branches are fully enumerated each one should be checked out into the environment directory, existing branches should be updated, and deleted branches should be removed.

**Comment by nvalentine-puppetlabs** [comment]
I concur this is the best approach possible given the how SVN "does" branches. Here's supporting docs from the most authoritative SVN book:

http://svnbook.red-bean.com/nightly/en/svn.branchmerge.using.html

I would add that to get as close to feature-parity as possible with the git backend of r10k we might want to build out 'tags' as well.

I don't envy you re implementing caching of the SVN repos. ;)

**Comment by adrienthebo** [comment]
I have a mostly complete implementation of this at https://github.com/adrienthebo/r10k/tree/feature/master/gh-104-svn-environments that can create environments and populate Puppetfile modules for each environment. There are a number of limitations such as prefixing isn't supported and environment names can't be sanitized. These limitations aside it's starting to approximate a final state so I need people to try it out, validate the behaviors, look for bugs, and help write documentation. ( @nvalentine-puppetlabs you asked for SVN support, so you better test it :P )

The given configuration is how I was testing this behavior:

``` yaml
---
sources:
  main:
    type: svn
    remote: 'https://github.com/jenkins-infra/jenkins-infra'
    basedir: '/home/adrien/development/r10k/testing/svn-env/environments'
```

**Comment by nvalentine-puppetlabs** [comment]
To be clear, _I_ would never ask for such a thing. But customers.... ;)

I will test it out a bit this week. Thanks for putting in all of the effort so far. :)

**Comment by nvalentine-puppetlabs** [comment]
Basic sanity check using this control repo passed: https://github.com/nrvale0/r10k-svn-testing-control

Got all of the environment dirs expected and the relevant modules were placed according to Puppetfile.

Noticed that the :cachedir: parameter had no effect (dir not populated) but perhaps that is by design given the way SVN handles branches/tags etc.

**Comment by adrienthebo** [comment]
That's expected; there's no sane way to handle caching with SVN. It'll never be possible to do so SVN will never be as fast as Git.

**Comment by adrienthebo** [comment]
Merged into 1.3.x in 82fec8d.
